### DISCUSSION TITLE
How to build a query with multi level relationship

### DISCUSSION THREAD

**Discussion by pk4all** [adonisjs/core#1571] [Help] [answered]
Hello !

I have a user model with relationship with education model

```javascript 
@hasOne(() => Education,{
    localKey: 'id',
    foreignKey: 'uid'
  })
  public education: HasOne<typeof Education>

await User.query().preload('education').where education is ABC
```
How to write query with where for find a specific education category.

**Comment by thetutlage** [comment, marked as answer]
Covered in the docs https://preview.adonisjs.com/guides/model-relations/introduction#relationship-constraints

**Reply by pk4all** [reply]
@thetutlage I want to find all user which education in "graduation"
I want to find like 4.1, we used 
```javascript
.whereHas('comments', (builder) => {
    builder.where('is_published', true)
  }, '>', 2)
```
I am not find like this is 5.

**Reply by thetutlage** [reply]
You mean, you want to limit the users result set based upon their education level?

**Reply by pk4all** [reply]
Yes right.

**Reply by thetutlage** [reply]
Also covered in the docs https://preview.adonisjs.com/guides/model-relations/introduction#querying-relationship-existence

**Reply by pk4all** [reply]
@thetutlage Thanks.
But I am getting an error => Property 'whereHas' does not exist on type 'ModelQueryBuilderContract<typeof User, User>'"

**Reply by thetutlage** [reply]
Make sure you are running latest version of lucid. Docs are always your best friend. https://preview.adonisjs.com/releases/lucid/version-8_3_2/

**Reply by pk4all** [reply]
@thetutlage Thanks.
I have installed 8.4.2. But I am getting a new error on on my model subquery 
```javascript
@hasMany(() => UserMasterGodown,{
    localKey: 'id',
    foreignKey: 'masterFor',
    onQuery: (query) => query.preload('master').where('masterType', 6)
  })
  public hashtags: HasMany<typeof UserMasterGodown>
```
error is =>

```
(property) ModelQueryBuilderContract<Model extends LucidModel, Result extends unknown = InstanceType<Model>>.preload: Preload<UserMasterGodown, RelationQueryBuilderContract<typeof UserMasterGodown, any>> | Preload<UserMasterGodown, RelationSubQueryBuilderContract<...>>
Define relationships to be preloaded

This expression is not callable.
  Each member of the union type 'Preload<UserMasterGodown, RelationQueryBuilderContract<typeof UserMasterGodown, any>> | Preload<UserMasterGodown, RelationSubQueryBuilderContract<...>>' has signatures, but none of those signatures are compatible with each other
```

**Reply by thetutlage** [reply]
Oh yes, you need to wrap your `preload` code inside an if statement.

```ts
onQuery: (query) => {
  if (!query.isRelatedSubQuery) {
    query.preload('master').where('masterType', 6)
  }
}
```

The `conditional` is required so that you are not preloading when your `onQuery` hook is executed for constructing a `whereHas` subquery. I will cover this properly in the docs

**Reply by pk4all** [reply]
@thetutlage Thanks a lot for such type quick support.
It's perfectly working for me.
